problem-spring-common/src/main/java/org/zalando/problem/spring/common/AdviceTrait.java

```diff
@@ -0,0 +1,141 @@
+ * use à la carte.
+ * </p>
+ * <p>
+ * Advice traits are grouped in packages, based on they use cases. Every package has a composite advice trait that
+ * bundles all traits of that package.
+ * </p>
+ *
+ * @see ControllerAdvice
+ * @see ExceptionHandler
+ * @see Throwable
+ * @see Exception
+ * @see Problem
+ */
+@API(status = STABLE)
+public interface AdviceTrait {
+
+    Logger LOG = LoggerFactory.getLogger(AdviceTrait.class);
+
+    default ThrowableProblem toProblem(final Throwable throwable) {
+        final StatusType status = Optional.ofNullable(resolveResponseStatus(throwable))
+                .<StatusType>map(ResponseStatusAdapter::new)
+                .orElse(Status.INTERNAL_SERVER_ERROR);
+
+        return toProblem(throwable, status);
+    }
+
+    @API(status = MAINTAINED)
+    default ResponseStatus resolveResponseStatus(final Throwable type) {
+        @Nullable final ResponseStatus candidate = findMergedAnnotation(type.getClass(), ResponseStatus.class);
+        return candidate == null && type.getCause() != null ? resolveResponseStatus(type.getCause()) : candidate;
+    }
+
+    default ThrowableProblem toProblem(final Throwable throwable, final StatusType status) {
+        return toProblem(throwable, status, Problem.DEFAULT_TYPE);
+    }
+
+    default ThrowableProblem toProblem(final Throwable throwable, final StatusType status, final URI type) {
+        final ThrowableProblem problem = prepare(throwable, status, type).build();
+        final StackTraceElement[] stackTrace = createStackTrace(throwable);
+        problem.setStackTrace(stackTrace);
+        return problem;
+    }
+
+    default ProblemBuilder prepare(final Throwable throwable, final StatusType status, final URI type) {
+        return Problem.builder()
+                .withType(type)
+                .withTitle(status.getReasonPhrase())
+                .withStatus(status)
+                .withDetail(throwable.getMessage())
+                .withCause(Optional.ofNullable(throwable.getCause())
+                    .filter(cause -> isCausalChainsEnabled())
+                    .map(this::toProblem)
+                    .orElse(null));
+    }
+
+    default StackTraceElement[] createStackTrace(final Throwable throwable) {
+        final Throwable cause = throwable.getCause();
+
+        if (cause == null || !isCausalChainsEnabled()) {
+            return throwable.getStackTrace();
+        } else {
+
+            final StackTraceElement[] next = cause.getStackTrace();
+            final StackTraceElement[] current = throwable.getStackTrace();
+
+            final int length = current.length - lengthOfTrailingPartialSubList(asList(next), asList(current));
+            final StackTraceElement[] stackTrace = new StackTraceElement[length];
+            System.arraycopy(current, 0, stackTrace, 0, length);
+            return stackTrace;
+        }
+    }
+
+    default boolean isCausalChainsEnabled() {
+        return false;
+    }
+
+    default void log(
```

Comment:
We would be missing the request in all of these here. I wanted to pass it everywhere so that implementations are free to use it to get access to headers or anything else. Trying to find a common part for mvc and webflux might not be possible for those methods.

Reply:
There's another `log` method in problem-spring-web which has the request param and defaults to calling this one. Do you feel that's too many methods ? I can make it a static function instead.

Reply:
We need another way to share this between mvc and webflux. An interface will expose it in both, but clients are not meant to see this one. Static methods are most likely our best option here.

Reply:
Changed to static